### DISCUSSION TITLE
[2023-06-26] GUI Team Demo Day

### DISCUSSION THREAD

**Discussion by wdanilo** [enso-org/enso#7126] [Weekly Demo Day]
## Overview

Last week, I was working on 2 things: integrating the new FRP Engine to Enso (2 days) and improved design of Component Browser and GUI experience (4 days). 

## FRP integration

| Task  | Progress 1 Week Ago | Progress Today | Days Needed To Finish It |
| ------------- | ------------- | ------------- |------------- |
| Moving old FRP nodes to new impl  | 66% (66/99)  | 100% (99/99)  | - |
| Testing all FRP nodes  | 0% (0/99)   | 10% (10/99)  | 1-2 days |
| Adding bridge network to the new impl | 0% |  0% | < 1 day |
| Moving `network::extend!` macro to new impl  | 0% | 10%  | 1 day |
| Heavy testing and debugging  | 0%   | 0%  | 1-2 days |

## The Component Browser Design
This task was not scheduled and I was initially thinking that it will be a small 1-day task to allow seamless AI integration with Enso, however, I was involved in multiple conversations with James and Ned about the current usability issues of component browser with very convincing arguments of why it is not where we want it to be, so a significant amount of work was needed to bring it to the state where it will solve these issues. We should think about it as design iteration – the previous design gave us a lot of thoughts and we were able to validate a lot of ideas. This improvement is already discussed with James, Ned, Adam, and Greg. You can find the detailed [description here](https://github.com/enso-org/design/pull/56). I gathered several comments about the design, I will apply them today and merge it tomorrow.

<img width="1437" alt="image" src="https://github.com/enso-org/design/assets/1623053/5840acdf-eb34-43f8-bdfa-dee7c04b95b5">

<img width="548" alt="image" src="https://github.com/enso-org/design/assets/1623053/3a9acf27-da21-474f-8536-3f94754f346e">

**Comment by vitvakatu** [comment]
- Fix for redundant imports in dropdown widget got merged: https://github.com/enso-org/enso/pull/7028
- Bumping `rustc` is almost done, but we faced build failure on Linux. The investigation is ongoing: https://github.com/enso-org/enso/pull/7109
- Resizing visualizations is in progress. The core of the implementation is ready, but more refactoring is on the way: https://github.com/enso-org/enso/issues/7048


https://github.com/enso-org/enso/assets/6566674/aa7ca5c9-add9-4391-942b-44d2f071eb1b

**Comment by mwu-tow** [comment]
CI — macOS signing and notarization was temporarily disabled. The issue has been addressed.
Observed relatively high instability, caused by the GH-side issues (though it was not a „big” failure). We should consider implementing tasks like https://github.com/enso-org/enso/issues/5236 to handle such cases better.

Remote logging — PR waits for review. https://github.com/enso-org/enso/pull/6582
Undo redo — debugging visualization-related issues. Split one simple improvement into a separate PR — https://github.com/enso-org/enso/pull/7121. Further investigation is needed.
Logging to file — delaying, as it is not a blocker and requires some more design work.

**Comment by MichaelMauderer** [comment]
Worked together with Jan on implementing a new AI completion searcher that can be used to protyle functionality related to generating searcher completions based on AI language models.

https://github.com/enso-org/enso/assets/1428930/1262cbdd-b5c9-4877-820d-3be1b56d7adc

**Comment by farmaazon** [comment]
I worked only 2 days last week. I fixed the problem with "flickering" node when connecting ports:

https://user-images.githubusercontent.com/3919101/247930326-112cc7be-eb31-444d-a0c8-d69c2013d47d.mp4

The flickering was caused by the fact the "Pending" updates from the engine overrode the method_call field with `None`, so we were removing gray ports only to re-instantiate them once the actual value arrived from the engine. Now, we do not remove method_call on Pending updates.

Besides, I was participating in discussions about AI integration and new design of things. Created a draft list of tasks needed for new design to work (at least visually): #7125

**Comment by kazcw** [comment]
I've been working on APIs supporting focus behavior, to implement https://github.com/enso-org/enso/issues/5337 and https://github.com/enso-org/enso/pull/7039#pullrequestreview-1483805352, fixing #6124. I have the implementation working, and will clean and document it today (so, the demo will be next week).

The main new API is a `focused_shortcuts` capability. This is quite different from the old approach of using a `focused` condition in a `default_shortcuts` rule. `focused_shortcuts` are propagated up the focus hierarchy. They are delivered to only the focused objects, rather than type-based dispatch. They propagate until something focused handles them (or until they reach the root and are checked against global rules); this allows a "more focused" descendant object to have the first chance to handle an event, before falling back to ancestors.

**Comment by Frizi** [comment]
- Finally finished edge refactor #7041, and fixed one regression that come from it #7102
- Resumed work on design changes. Refactored the old branch to latest develop and continuing with it. Added icon display to the nodes.
<img width="548" alt="image" src="https://github.com/enso-org/enso/assets/919491/6c0b270a-754f-4e30-88fa-14dc750a0ea8">

- Changed how icons can be defined in the libraries, and modified the "icons" example scene to make it useful for the Libs team. #7123
- Attempted to fold component group data into suggestion database, and understood why it is currently not really possible. Started discussion on potential changes to the group definition mechanism, such that it will be easier to define and use on the IDE side for node coloring purposes. https://github.com/orgs/enso-org/discussions/7120

**Comment by JaroslavTulach** [comment]
Engine related performance work:
- #7101
- #7097
    -   Óda for throttling: https://github.com/enso-org/enso/issues/7087#issuecomment-1600072772
- #7116
